### ISSUE TITLE
AssertionException when using the Statistics service

### ISSUE THREAD

**Issue by DavideD** [hibernate/hibernate-reactive#1015]
This error occurs when I run [this test](4897fbd45f5f26463286d48739136a58c462f0e5):

```
    java.lang.AssertionError: Nested calls to partialFlushStart                                 
        at org.hibernate.engine.internal.StatisticalLoggingSessionEventListener.partialFlushStart(StatisticalLoggingSessionEventListener.java:241)                                                                   
        at org.hibernate.engine.internal.SessionEventListenerManagerImpl.partialFlushStart(SessionEventListenerManagerImpl.java:239)
        at org.hibernate.reactive.event.impl.DefaultReactiveAutoFlushEventListener.reactiveOnAutoFlush(DefaultReactiveAutoFlushEventListener.java:39)                                                                
        at org.hibernate.event.service.internal.EventListenerGroupImpl.lambda$fireEventOnEachListener$0(EventListenerGroupImpl.java:133)
        at java.base/java.util.concurrent.CompletableFuture.uniComposeStage(CompletableFuture.java:1106)                                                                                                             
        at java.base/java.util.concurrent.CompletableFuture.thenCompose(CompletableFuture.java:2235)
        at java.base/java.util.concurrent.CompletableFuture.thenCompose(CompletableFuture.java:143)                                                                                                                  
        at org.hibernate.event.service.internal.EventListenerGroupImpl.fireEventOnEachListener(EventListenerGroupImpl.java:133)                      
```

This is the testcase for `CachedQueryResultsTest`:
```
	@Test
	public void testQueryPlanCacheHits(TestContext context) {
		test( context, criteriaFindAll()
				.call( CachedQueryResultsTest::criteriaFindAll )
				.call( CachedQueryResultsTest::criteriaFindAll )
				.invoke( () -> context.assertEquals( 2L, statistics().getQueryPlanCacheHitCount() ) )
		);
	}

	private static StatisticsImplementor statistics() {
		return ( (Implementor) getSessionFactory() ).getServiceRegistry().getService(
				StatisticsImplementor.class );
	}

	private static Uni<List<Fruit>> criteriaFindAll() {
		final Mutiny.SessionFactory sf = getMutinySessionFactory();
		return sf.withStatelessSession( s -> s.createQuery( criteriaQuery( sf.getCriteriaBuilder() ) ).getResultList()
		);
	}

	private static CriteriaQuery<Fruit> criteriaQuery(CriteriaBuilder criteriaBuilder) {
		CriteriaQuery<Fruit> criteriaQuery = criteriaBuilder.createQuery( Fruit.class );
		Root<Fruit> from = criteriaQuery.from( Fruit.class );
		criteriaQuery.select( from );
		return criteriaQuery;
	}
```

```

**Comment by blafond** [comment]
@DavideD @gavinking 

Revisted this today and the error has changed to: 
```
    java.lang.AssertionError: Not equals : 2 != 0
        at io.vertx.ext.unit.impl.TestContextImpl.reportAssertionError(TestContextImpl.java:362)
        at io.vertx.ext.unit.impl.TestContextImpl.assertEquals(TestContextImpl.java:258)
        at io.vertx.ext.unit.impl.TestContextImpl.assertEquals(TestContextImpl.java:245)
        at org.hibernate.reactive.CachedQueryResultsTest.lambda$testQueryPlanCacheHits$10(CachedQueryResultsTest.java:182)
```
I'm new to this feature ... and the errors are drastically different.   Any suggestions on where to look?

**Comment by DavideD** [comment]
I guess it works now because we have enabled the statistics. But it seems we don't collect this value correctly?
As usual, you would need to check what happens in ORM and see that we do the same in Hibernate Reactive.

**Comment by DavideD** [comment]
When I enable the statistics, this two tests fail:

* testLoadFromCachedQueryResultAndRegularQuery
* testLoadFromCachedQueryResultAndNamedQuery

**Comment by blafond** [comment]
I repeated your failures after adding that config property.

I enabled the commented out code in [ReactiveSessionImpl](https://github.com/hibernate/hibernate-reactive/blob/ad5565ab34d112000840bc9f782b6dc84cb52f60/hibernate-reactive-core/src/main/java/org/hibernate/reactive/session/impl/ReactiveSessionImpl.java#L410-L413) and all tests in `CachedQueryResultsTest` succeed.

However, a number of different tests failed when running a full build.

So there's an interaction between flushing & the transaction boundaries that I don't comprehend yet.
